server/src/main/java/org/elasticsearch/cluster/coordination/PreVoteCollectorFactory.java

```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    // package-private for testing, actual usages TODO
+    long getMaxTermSeen() {
+        return maxTermSeen.get();
+    }
+
+    @Override
+    public String toString() {
+        return "PreVoteCollectorFactory{" +
+            "currentLeader=" + currentLeader +
+            ", preVoteResponse=" + preVoteResponse +
+            '}';
+    }
+
+    private class PreVotingRound implements Releasable {
+        private final Set<DiscoveryNode> preVotesReceived = newConcurrentSet();
+        private final AtomicBoolean electionStarted = new AtomicBoolean();
+        private final PreVoteRequest preVoteRequest;
+        private final ClusterState clusterState;
+        private final AtomicBoolean isRunning = new AtomicBoolean();
+
+        PreVotingRound(ClusterState clusterState) {
+            this.clusterState = clusterState;
+            preVoteRequest = new PreVoteRequest(transportService.getLocalNode(), preVoteResponse.getCurrentTerm());
+        }
+
+        void start(final Iterable<DiscoveryNode> broadcastNodes) {
+
+            final boolean isRunningChanged = isRunning.compareAndSet(false, true);
+            assert isRunningChanged;
+
+            broadcastNodes.forEach(n -> transportService.sendRequest(n, REQUEST_PRE_VOTE_ACTION_NAME, preVoteRequest,
+                new TransportResponseHandler<PreVoteResponse>() {
+                    @Override
+                    public void handleResponse(PreVoteResponse response) {
+                        handlePreVoteResponse(response, n);
+                    }
+
+                    @Override
+                    public void handleException(TransportException exp) {
+                        if (exp.getRootCause() instanceof CoordinationStateRejectedException) {
+                            logger.debug("{} failed: {}", this, exp.getRootCause().getMessage());
+                        } else {
+                            logger.debug(new ParameterizedMessage("{} failed", this), exp);
+                        }
+                    }
+
+                    @Override
+                    public String executor() {
+                        return Names.GENERIC;
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "TransportResponseHandler{" + PreVoteCollectorFactory.this + ", node=" + n + '}';
+                    }
+                }));
+        }
+
+        private void handlePreVoteResponse(PreVoteResponse response, DiscoveryNode sender) {
+            if (isRunning.get() == false) {
+                logger.debug("{} ignoring {} from {}, no longer running", this, response, sender);
+                return;
+            }
+
+            final long currentMaxTermSeen = updateMaxTermSeen(response.getCurrentTerm());
+
+            final PreVoteResponse currentPreVoteResponse = preVoteResponse;
```

Comment:
I wonder if we should capture this in the PreVotingRound constructor. This should very rarely be changing (similarly to cluster state) if there is an ongoing prevoting round. This makes this instance also internally more consistent. Finally, I would like to see if we can make this inner class static, so that we know exactly about all the dependencies it has.

Reply:
We can use the current (i.e. last-accepted) cluster state for this, I think.

The inner class uses these from the outer class:

- `logger`
- `transportService`
- `startElection.run()`
- `toString()`
- `preVoteResponse.getCurrentTerm()` (in the constructor)

I got rid of the `getCurrentTerm()` call in the constructor, but I would prefer to use the others as-is rather than making it static.